src/librbd/Utils.h

```diff
@@ -154,14 +154,24 @@ std::string unique_lock_name(const std::string &name, void *address);
 
 template <typename I>
 std::string data_object_name(I* image_ctx, uint64_t object_no) {
-  char buf[RBD_MAX_OBJ_NAME_SIZE];
-  size_t length = snprintf(buf, RBD_MAX_OBJ_NAME_SIZE,
-                           image_ctx->format_string, object_no);
-  ceph_assert(length < RBD_MAX_OBJ_NAME_SIZE);
-
-  std::string oid;
-  oid.reserve(RBD_MAX_OBJ_NAME_SIZE);
-  oid.append(buf, length);
+  size_t object_no_width = image_ctx->old_format ? 12 : 16;
+  auto prefix_length = image_ctx->object_prefix.length();
+  std::string oid(prefix_length + 1 + object_no_width, '0');
+  auto buf = oid.data();
```

Comment:
Nit: I'd make `object_no_width` and `prefix_length` an int.  "int by default" is a good rule to follow and you made `pos` which is derived from them an int ;)